doc/source/development/rfc/rfc89_sql_logging_callback.rst

```diff
@@ -0,0 +1,116 @@
+.. _rfc-89:
+
+=============================================================
+RFC 89: SQL logging callback
+=============================================================
+
+============== =============================================
+Author:        Alessandro Pasotti
+Contact:       elpaso @ itopen.it
+Started:       2022-Nov-30
+Status:        Draft
+Target:        GDAL 3.7
+============== =============================================
+
+Summary
+-------
+
+The document proposes and describes the introduction of a new
+callback function to allow applications that use the GDAL library
+to monitor the SQL queries actually sent from GDAL to the database.
+
+Motivation
+----------
+
+Applications (for example QGIS) may provide a debugging panel that
+allow users to monitor the SQL command that are sent to the layers
+stored on databases.
+
+This panel implemented in QGIS is very useful to identify erroneous
+or inefficient queries and also made the users more conscious about
+what is going on in the application when layers are loaded and features
+are fetched.
+
+
+Technical details
+-----------------
+
+The implementation details are still to be defined, a possible
+implementation would provide a method to `GDALDataset` that
+allows client code to set the callback function (with some opaque
+context to be passed back to client code).
+
+The callback function will then be called from the drivers that
+make database calls each time a SQL command is sent to the backend,
+in case of prepared queries the actual SQL after parameter
+substitutions should be sent.
+
+The callback will provide additional information about the executed
+query (if available):
+
+- error string message
+- number of affected/retrieved records
+- time taken to execute the query
+
```

Comment:
It should be documented that the callback may be called by multiple threads in a concurrent way, and that the implementation of the callback should be robust to that (use of lock typically). That may arise in drivers that use multi-threading as an optimization implementation detail. And this is typically the case of the GeoPackage driver in its Arrow stream interface. Cf https://github.com/OSGeo/gdal/blob/master/ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp#L7099.

Regarding the implementation in the GPKG driver, it should around https://github.com/OSGeo/gdal/blob/master/ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp#L7121 install the callback set initially on the main dataset to the ``task->m_poDS`` internal dataset that will run in a dedicated thread.

Reply:
I've shamelessly copy-pasted your sentence.

Reply:
About the implementation, I've installed the callback this way: https://github.com/elpaso/gdal/commit/4b1bf394e5af24e7758aa31c7162ee61f9a768dd#diff-125788fb2cb0d32e396c0af12cb5bbc19b12a006603dddc29236a3afdb9c8096R4342

Reply:
> I've installed the callback this way

My point regarding the multi-threaded optimization was that, to instrument everything, the callback also needs to be propagated by the driver itself from the main dataset to the auxiliary datasets it opens on the same file, and that should be done around https://github.com/OSGeo/gdal/blob/master/ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp#L7121

Reply:
See https://github.com/elpaso/gdal/commit/4e3f0e8eb742599320ba9e1da4eb4ae988004a77